src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1661,18 +1661,15 @@ abstract class RefChecks extends Transform {
         case TypeApply(fun, targs) =>
           isClassTypeAccessible(fun)
         case Select(module, apply) =>
-          ( // scala/bug#4859 `CaseClass1().InnerCaseClass2()` must not be rewritten to `new InnerCaseClass2()`;
-            //          {expr; Outer}.Inner() must not be rewritten to `new Outer.Inner()`.
-            treeInfo.isQualifierSafeToElide(module) &&
-              // scala/bug#5626 Classes in refinement types cannot be constructed with `new`. In this case,
-              // the companion class is actually not a ClassSymbol, but a reference to an abstract type.
-              module.symbol.companionClass.isClass
-            )
+          // scala/bug#4859 `CaseClass1().InnerCaseClass2()` must not be rewritten to `new InnerCaseClass2()`;
+          //          {expr; Outer}.Inner() must not be rewritten to `new Outer.Inner()`.
+          treeInfo.isQualifierSafeToElide(module) &&
+          // scala/bug#5626 Classes in refinement types cannot be constructed with `new`.
+          !module.exists { case t @ Select(_, _) => t.symbol != null && t.symbol.isStructuralRefinementMember case _ => false }
         case x => throw new MatchError(x)
       }
-
       sym.name == nme.apply &&
-        !(sym hasFlag STABLE) && // ???
+        !sym.hasStableFlag && // ???
```

Comment:
This goes back to the introduction of `isSourceMethod` in 2005, which (as now commented) uses stable flag to distinguish accessors (including what we call case accessors) from regular methods. The world was much simpler back then.